src/reconcile/rec_write.c

```diff
@@ -1290,8 +1309,8 @@ __rec_txn_read(WT_SESSION_IMPL *session, WT_RECONCILE *r,
 		}
```

Comment:
I think there's an additional problem in the checkpoint loop. The `__wt_txn_upd_visible` check doesn't include any check for `upd->txnid == WT_UPDATE_RESERVED`, which I think means we can legitimately return visibility for a reserved update, again leading to assertion failure at line 1312.

We could do:
```
if (upd->type == WT_UPDATE_RESERVED || txnid == WT_TXN_ABORTED)
```
or:
```
if (upd->type != WT_UPDATE_RESERVED && __wt_txn_upd_visible(session, upd))
```
so it looks like `__wt_txn_read()`.